pkg/client/userd/commands/commands.go

```diff
@@ -26,22 +29,93 @@ func GetCwd(ctx context.Context) string {
 	return ""
 }
 
+type command interface {
+	init(context.Context)
+	cobraCommand(context.Context) *cobra.Command
+	group() string
+}
+
+func commands() []command {
+	return []command{
+		&interceptCommand{},
+	}
+}
+
 // GetCommands will return all commands implemented by the connector daemon.
-func GetCommands() cliutil.CommandGroups {
-	return cliutil.CommandGroups{}
+func GetCommands(ctx context.Context) cliutil.CommandGroups {
+	var groups = cliutil.CommandGroups{}
+	for _, cmd := range commands() {
+		var (
+			groupName = cmd.group()
+			group     = groups[groupName]
+		)
+		cmd.init(ctx)
+		groups[groupName] = append(group, cmd.cobraCommand(ctx))
+	}
+	return groups
 }
 
 // GetCommandsForLocal will return the same commands as GetCommands but in a non-runnable state that reports
 // the error given. Should be used to build help strings even if it's not possible to connect to the connector daemon.
-func GetCommandsForLocal(err error) cliutil.CommandGroups {
-	groups := GetCommands()
-	for _, cmds := range groups {
-		for _, cmd := range cmds {
-			cmd.RunE = func(_ *cobra.Command, _ []string) error {
-				// err here will be ErrNoUserDaemon "telepresence user daemon is not running"
-				return fmt.Errorf("unable to run command: %w", err)
-			}
+func GetCommandsForLocal(ctx context.Context, err error) cliutil.CommandGroups {
+	var groups = cliutil.CommandGroups{}
+	for _, cmd := range commands() {
+		var (
+			groupName = cmd.group()
+			group     = groups[groupName]
+			cc        = cmd.cobraCommand(ctx)
+		)
+		cc.RunE = func(_ *cobra.Command, _ []string) error {
+			// err here will be ErrNoUserDaemon "telepresence user daemon is not running"
+			return fmt.Errorf("unable to run command: %w", err)
 		}
+		groups[groupName] = append(group, cc)
 	}
 	return groups
 }
+
+type sessKey struct{}
+
+func WithSession(ctx context.Context, s trafficmgr.Session) context.Context {
+	return context.WithValue(ctx, sessKey{}, s)
+}
+
+func GetSession(ctx context.Context) trafficmgr.Session {
+	s, _ := ctx.Value(sessKey{}).(trafficmgr.Session)
+	return s
+}
+
+type connectorKey struct{}
+
+func WithConnectorServer(ctx context.Context, cs connector.ConnectorServer) context.Context {
+	return context.WithValue(ctx, connectorKey{}, cs)
+}
+
+func GetConnectorServer(ctx context.Context) connector.ConnectorServer {
+	cs, _ := ctx.Value(connectorKey{}).(connector.ConnectorServer)
+	return cs
+}
+
+type ctxCancellationHandlerFuncKey struct{}
+
+func WithCtxCancellationHandlerFunc(ctx context.Context) context.Context {
+	var f func()
+	return context.WithValue(ctx, ctxCancellationHandlerFuncKey{}, &f)
+}
+
+func SetCtxCancellationHandlerFunc(ctx context.Context, f func()) {
+	fp, ok := ctx.Value(ctxCancellationHandlerFuncKey{}).(*func())
+	if !ok {
+		return
+	}
+	*fp = f
+}
+
+func GetCtxCancellationHandlerFunc(ctx context.Context) func() {
+	fp, ok := ctx.Value(ctxCancellationHandlerFuncKey{}).(*func())
+	if !ok {
+		return nil
+	}
+
+	return *fp
+}
```

Comment:
future consideration but we may wanna create some sort of command context struct to store all this stuf